Clarify Disable Autolab Submissions Option

jhs-panda opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description
<!--- Describe your changes in detail -->
Adjusted UI to make the "Disable Autolab submissions" button and functionality more noticeable. Also edited description on submissions page.
Before:
<img width="1054" alt="Screen Shot 2024-09-27 at 4 41 53 PM" src="https://github.com/user-attachments/assets/6d773e0b-8cde-4a21-8f8b-112a6c9a895d">
<img width="827" alt="Screen Shot 2024-09-27 at 4 46 38 PM" src="https://github.com/user-attachments/assets/931bc48a-305d-40e5-a273-ffe86bae87f0">

After:
<img width="1069" alt="Screen Shot 2024-09-27 at 4 44 31 PM" src="https://github.com/user-attachments/assets/86ce7d99-c2b5-41d4-9a54-8604cc4918fc">
<img width="1045" alt="Screen Shot 2024-09-27 at 4 44 43 PM" src="https://github.com/user-attachments/assets/3209cbdd-96c5-4e64-9e71-76d29dd6063e">
<img width="1322" alt="image" src="https://github.com/user-attachments/assets/e449878f-6d53-408e-839c-c1ca6ba0e3ba">
Also added some error checking for the handin filename:
<img width="1029" alt="image" src="https://github.com/user-attachments/assets/6808141d-cbaa-45c9-b388-c107444d1dd4">

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
Some classes have assignments that shouldn't be submitted through Autolab. The functionality for disabling submissions through Autolab exists, but it is currently a bit unclear what the button does and many classes therefore don't use it. Fixes #2216.

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->
Tested locally (see screenshots) to make sure changes were saved and toggle between disable/enable handins and gray-out worked as expected.

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
- [x] I have run rubocop and erblint for style check. If you haven't, run `overcommit --install && overcommit --sign` to use pre-commit hook for linting
- [ ] My change requires a change to the documentation, which is located at [Autolab Docs](https://docs.autolabproject.com/)
- [ ] I have updated the documentation accordingly, included in this PR

## Other issues / help required
<!--- Do you have any other relevant issues / questions? --->
<!--- For example, if you require assistance for a certain part of your PR --->
<!--- or are facing specific issues while creating the pull request that you would like to highlight --->

evanyeyeye reviewed (changes requested):
Mostly looks good!

However, this PR forces the "Handin filename" field to always be `handin.c` on the frontend view, even when it has been successfully modified.

Also, I noticed that in your "After" screenshot the "Autograder settings" button is gone in the left navigation column. Is that an effect of the PR?

jhs-panda commented:
> However, this PR forces the "Handin filename" field to always be `handin.c` on the frontend view, even when it has been successfully modified.

I believe I have corrected this my recent commits.

> Also, I noticed that in your "After" screenshot the "Autograder settings" button is gone in the left navigation column. Is that an effect of the PR?

Sorry, I chose different assessments pages to screenshot and one of them wasn't an autograded assignment. I've corrected the image in the "After" screenshot.

Thanks!

jhs-panda commented on app/views/assessments/_edit_handin.html.erb:
```diff
@@ -29,14 +46,17 @@
+                 value: f.object.handin_filename.presence || "handin.c",
+                 help_text: "The suffix that is appended to student submission files. Autolab stores submission files in the handin directory as email/version_fname",
+                 placeholder: "E.g. mm.c",
+                 class: "handin-filename-field" %>
```
Functionality doesn't work, tried.

jhs-panda replied:
Took the suggestion for the most part, excluded hyphens because its inclusion allowed other invalid characters to be used without flashing error.

jhs-panda reviewed (commented):
Processed suggested edits and CodeRabbit suggestions.

jhs-panda commented on app/views/assessments/_edit_handin.html.erb:
```diff
@@ -29,14 +45,22 @@
+                help_text: "Check this to disallow handins through Autolab. This option can be used to track scores for assignments that are not submitted through Autolab such as midterms and written assignments.",
+                class: "disable-handins-toggle",
+                onchange: "toggleHandinAutolab(this)" %>
+<%= f.text_field :handin_filename,
+                 value: f.object.handin_filename.presence,
+                 help_text: "The suffix that is appended to student submission files. Autolab stores submission files in the handin directory as email/version_fname",
+                 placeholder: "E.g. mm.c",
+                 class: "handin-filename-field",
+                 pattern: "^[a-zA-Z0-9._-]+$",
+                 title: "Only alphanumeric characters, dots, hyphens, and underscores are allowed." %>
```
We want to include hyphens.

jhs-panda commented:
> Also, the sanitization for handin filename is probably a little too restrictive. For example, hyphens are used pretty regularly for 15213 assessments.
For some reason, allowing hyphens lets any type of character in the filename (adding hyphens was suggested by CodeRabbit before and any filename was allowed). That's why I removed it and allowed the other characters. If we want it to be on the profs' side to have valid filenames, it's fine. Otherwise, allowing hyphens in the pattern will mess up any filename error checking.

evanyeyeye reviewed (changes requested):
The edge case fix seems to work, but its a bit hacky. Not sure if this is the best way to do things, but I think it suffices for now. Regarding the hyphen, it will work if you escape it with two backslash characters. However, I'm not sure if we can fully account for all handin filenames used by current courses. Maybe we want to give instructors more freedom on this.

evanyeyeye commented:
LGTM.

evanyeyeye reviewed (approved):
LGTM!
